File: lc/T401-450/t_416.java
package leetcoder;


import java.util.*;


class Solution {
    public boolean canPartition(int[] nums) {
        int  sum = 0;
        for (int num : nums) {
            sum+=num;
        }
        if((sum&1) ==1) return false;
        sum>>>=1;
        boolean [] dp = new boolean[sum+1];
        dp[0] = true;
        for (int num : nums) {
            for (int i = sum; i >= num; i--) {
                dp[i] = dp[i] | (dp[i -num]);
            }
        }
        return dp[sum];

    }
}

File: leetcode/lc/T451-500/T416.cc
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        auto  sum = accumulate(nums.begin(), nums.end(),0);
        if(sum&1){
            return false;
        }
        sum>>=1;
        vector<int> dp(sum+1,0);
        dp[0]= 1;
        for(auto it :nums){
            for(int j=sum;j>=1;j--){
                if(j-it>=0 && dp[j-it]>0){
                    dp[j] = 1;
                }
            }
        }
        return dp[sum];

    }
};